File: frontend/src/app/core/marketplace.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from './api.service';
import {
  FulfillmentType,
  Order,
  OrderStatus,
  Product,
  ProductCategory,
  ProductUnit,
  Review,
} from './models';

@Injectable({ providedIn: 'root' })
export class MarketplaceService {
  constructor(private api: ApiService) {}

  searchProducts(params: {
    lat?: number;
    lng?: number;
    radiusKm?: number;
    category?: ProductCategory;
    q?: string;
  }): Promise<Product[]> {
    const query = new URLSearchParams();
    if (params.lat != null) query.set('lat', String(params.lat));
    if (params.lng != null) query.set('lng', String(params.lng));
    if (params.radiusKm != null) query.set('radiusKm', String(params.radiusKm));
    if (params.category) query.set('category', params.category);
    if (params.q) query.set('q', params.q);
    const qs = query.toString();
    return this.api.request<Product[]>(`/api/products${qs ? `?${qs}` : ''}`, {}, false);
  }

  getProduct(id: string, lat?: number, lng?: number): Promise<Product> {
    const query = new URLSearchParams();
    if (lat != null) query.set('lat', String(lat));
    if (lng != null) query.set('lng', String(lng));
    const qs = query.toString();
    return this.api.request<Product>(`/api/products/${id}${qs ? `?${qs}` : ''}`, {}, false);
  }

  myProducts(): Promise<Product[]> {
    return this.api.request<Product[]>('/api/producer/products');
  }

  createProduct(payload: {
    name: string;
    description?: string;
    category: ProductCategory;
    price: number;
    unit: ProductUnit;
    quantity: number;
    latitude: number;
    longitude: number;
    locationLabel?: string;
    pickupAvailable?: boolean;
    deliveryAvailable?: boolean;
    photoUrls?: string[];
  }): Promise<Product> {
    return this.api.request<Product>('/api/producer/products', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  }

  deleteProduct(id: string): Promise<void> {
    return this.api.request<void>(`/api/producer/products/${id}`, { method: 'DELETE' });
  }

  createOrder(payload: {
    productId: string;
    quantity: number;
    fulfillment: FulfillmentType;
    message?: string;
  }): Promise<Order> {
    return this.api.request<Order>('/api/orders', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  }

  myOrders(): Promise<Order[]> {
    return this.api.request<Order[]>('/api/orders/mine');
  }

  producerOrders(): Promise<Order[]> {
    return this.api.request<Order[]>('/api/producer/orders');
  }

  updateOrderStatus(id: string, status: OrderStatus): Promise<Order> {
    return this.api.request<Order>(`/api/orders/${id}/status`, {
      method: 'PATCH',
      body: JSON.stringify({ status }),
    });
  }

  createReview(payload: { orderId: string; rating: number; comment?: string }): Promise<Review> {
    return this.api.request<Review>('/api/reviews', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  }

  producerReviews(producerId: string): Promise<Review[]> {
    return this.api.request<Review[]>(`/api/producers/${producerId}/reviews`, {}, false);
  }

  async uploadImage(file: File): Promise<string> {
    const form = new FormData();
    form.append('file', file);
    const res = await this.api.request<{ url: string }>('/api/uploads', {
      method: 'POST',
      body: form,
    });
    return res.url;
  }
}

File: mobile/lib/services/marketplace_service.dart
import 'package:lokal/models/models.dart';
import 'package:lokal/services/api_client.dart';

class MarketplaceService {
  MarketplaceService(this._api);

  final ApiClient _api;

  Future<List<Product>> searchProducts({
    double? lat,
    double? lng,
    int? radiusKm,
    ProductCategory? category,
    String? q,
  }) async {
    final query = <String, String>{
      if (lat != null) 'lat': '$lat',
      if (lng != null) 'lng': '$lng',
      if (radiusKm != null) 'radiusKm': '$radiusKm',
      if (category != null) 'category': categoryApi(category),
      if (q != null && q.isNotEmpty) 'q': q,
    };
    final list = await _api.request<List<dynamic>>(
      'GET',
      '/api/products',
      query: query,
      auth: false,
    );
    return list.map((e) => Product.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<Product> getProduct(String id, {double? lat, double? lng}) async {
    final query = <String, String>{
      if (lat != null) 'lat': '$lat',
      if (lng != null) 'lng': '$lng',
    };
    final json = await _api.request<Map<String, dynamic>>(
      'GET',
      '/api/products/$id',
      query: query,
      auth: false,
    );
    return Product.fromJson(json);
  }

  Future<List<Product>> myProducts() async {
    final list = await _api.request<List<dynamic>>('GET', '/api/producer/products');
    return list.map((e) => Product.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<Product> createProduct(Map<String, dynamic> body) async {
    final json = await _api.request<Map<String, dynamic>>(
      'POST',
      '/api/producer/products',
      body: body,
    );
    return Product.fromJson(json);
  }

  Future<void> deleteProduct(String id) async {
    await _api.request<void>('DELETE', '/api/producer/products/$id');
  }

  Future<Order> createOrder(Map<String, dynamic> body) async {
    final json = await _api.request<Map<String, dynamic>>('POST', '/api/orders', body: body);
    return Order.fromJson(json);
  }

  Future<List<Order>> myOrders() async {
    final list = await _api.request<List<dynamic>>('GET', '/api/orders/mine');
    return list.map((e) => Order.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<List<Order>> producerOrders() async {
    final list = await _api.request<List<dynamic>>('GET', '/api/producer/orders');
    return list.map((e) => Order.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<Order> updateOrderStatus(String id, OrderStatus status) async {
    final json = await _api.request<Map<String, dynamic>>(
      'PATCH',
      '/api/orders/$id/status',
      body: {'status': orderStatusApi(status)},
    );
    return Order.fromJson(json);
  }

  Future<void> createReview({
    required String orderId,
    required int rating,
    String? comment,
  }) async {
    await _api.request<Map<String, dynamic>>(
      'POST',
      '/api/reviews',
      body: {'orderId': orderId, 'rating': rating, 'comment': comment},
    );
  }

  Future<List<Review>> producerReviews(String producerId) async {
    final list = await _api.request<List<dynamic>>(
      'GET',
      '/api/producers/$producerId/reviews',
      auth: false,
    );
    return list.map((e) => Review.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<String> uploadImage(
    List<int> bytes, {
    required String filename,
    String? mimeType,
  }) =>
      _api.uploadImage(bytes, filename: filename, mimeType: mimeType);
}
